ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/slot/SlotStore.java

```diff
@@ -0,0 +1,222 @@
+    }
+
+    /**
+     * Retrieve the serialized state for an entry.
+     *
+     * @param key The unique identifier for the entry
+     * @return The serialized state
+     * @throws IOException if the entry is not found
+     */
+    public InputObjectState read(SlotStoreKey key) throws IOException {
+
+        Integer slotId = slotIdIndex.get(key);
+        if (slotId == null) {
+            throw new IOException("record not found for " + key);
+        }
+        byte[] data = slots.read(slotId);
+        // it's possible, though unlikely, for the data to be null here due to a concurrent remove
+        if (data == null) {
+            throw new IOException("record not found for " + key);
+        }
+        InputBuffer inputBuffer = new InputBuffer(data);
+        SlotStoreKey.unpackFrom(inputBuffer);
+
+        InputObjectState inputObjectState = new InputObjectState();
+        inputObjectState.unpackFrom(inputBuffer);
+
+        return inputObjectState;
+    }
+
+    /**
+     * Remove the state for an entry, freeing the slot.
+     * Depending on the configuration, this change may not be immediately persistent.
+     *
+     * @param key The unique identifier for the entry
+     * @return true on success, false otherwise
+     * @throws IOException unused for now in this impl.
+     */
+    public boolean remove(SlotStoreKey key) throws IOException {
+
+        Integer slotId = slotIdIndex.remove(key);
+        if (slotId == null) {
+            return false;
+        }
+
+        slots.clear(slotId, config.isSyncDeletes());
+
+        freeList.add(slotId);
+
+        return true;
+    }
+
+    /**
+     * Write (or overwrite) an entry with the given key and value.
+     *
+     * @param key               The unique identifier for the entry
+     * @param outputObjectState The serialized state
+     * @return true on success, false otherwise e.g. when the store is full.
+     * @throws IOException if serialization fails
+     */
+    public boolean write(SlotStoreKey key, OutputObjectState outputObjectState) throws IOException {
+
+        OutputObjectState record = new OutputObjectState();
+        key.packInto(record);
+        outputObjectState.packInto(record);
+        byte[] data = outputObjectState.buffer();
+
+        if (data.length > config.getBytesPerSlot()) {
+            throw new IOException("data too big for slot");
+        }
+
+        // We always write to a new slot, as overwrite in place may be non-atomic and
+        // risks leaving us with neither the before or after version for crash recovery.
+
+        Integer slotId = freeList.poll();
+        if (slotId == null) {
+            return false;
+        }
+
+        slots.write(slotId, data, config.isSyncWrites());
+        slotIdIndex.put(key, slotId);
+
+        Integer previousSlot = slotIdIndex.put(key, slotId);
+
+        // If it's a rewrite, we need to release the older version's slot
+        if (previousSlot != null) {
```

Comment:
So now there are two entries for the same state. Can periodic recovery kick in at this point and attempt to recover both?

Reply:
There is a time window where internally that's true, just as in wider recovery we may still have a log for a tx that has actually completed and isn't in the resource managers any more.
In the case of SlotStore it is (will be, when I write it...) indeterminate which state is used, which is fine so log as exactly one is. That's taken care of by the index, which only ever holds one even if two underlying slots are used. Periodic Recovery (in the ArjunaCore sense) sees only one, as it's going via the index. Recovery internally at the SlotStore level (one-off at startup time) will see both and insert only one into the index.

Reply:
Thanks for a thorough answer.